src/pymorize/global_attributes.py

```diff
@@ -0,0 +1,273 @@
+        license_text = self.cv["license"]["license"]
+        # make placeholders in license text
+        license_text = re.sub(r"<.*?>", "{}", license_text)
+        further_info_url = attrs_map_on_rule.get("further_info_url", None)
+        if further_info_url is None:
+            license_text = re.sub(r"\[.*?\]", "", license_text)
+            license_text = license_text.format(
+                selected_institution_id, license_id, license_url
+            )
+        else:
+            license_text = license_text.format(
+                selected_institution_id, license_id, license_url, further_info_url
+            )
+        grid_label = attrs_map_on_rule.get("grid_label", None)
+        if grid_label is None:
+            raise ValueError("Missing required attribute `grid_label`")
+        return {
+            "source_id": source_id,
+            "source": f"{model_component} ({cv['release_year']})",
+            "institution_id": selected_institution_id,
+            "institution": self.cv["institution_id"][selected_institution_id],
+            "grid": grid_description,
+            "grid_label": grid_label,
+            "nominal_resolution": native_nominal_resolution,
+            "license": license_text,
+        }
+
+    def _experiment_id_related(self, attrs_map_on_rule: dict) -> dict:
+        exp_id = attrs_map_on_rule.get("experiment_id")
+        cv = self.cv["experiment_id"][exp_id]
+        _activity_id = attrs_map_on_rule.get("activity_id", None)
+        activity_id = cv["activity_id"]
+        if _activity_id:
+            assert _activity_id in activity_id
+        else:
+            if len(activity_id) > 1:
+                raise ValueError(f"Mutiple activity_id found {activity_id}")
+            _activity_id = next(iter(activity_id))
+        sub_experiment_id = " ".join(cv["sub_experiment_id"])
+        if sub_experiment_id == "none":
+            sub_experiment = "none"
+        else:
+            sub_experiment = sub_experiment_id.split()[0]
+        return {
+            "activity_id": _activity_id,
+            "experiment_id": exp_id,
+            "experiment": cv["experiment"],
+            "sub_experiment": sub_experiment,
+            "sub_experiment_id": sub_experiment_id,
+            "source_type": " ".join(cv["required_model_components"]),
+        }
+
+    def _header_related(self, attrs_map_on_rule: dict) -> dict:
+        """
+        Extracts header related global attributes from a DataRequestRule.
+
+        Parameters
+        ----------
+        rule : DataRequestRule
+            The data request rule to extract the header related global
+            attributes from.
+
+        Returns
+        -------
+        dict
+            A dictionary of global attributes
+        """
+        d = {}
+        header = attrs_map_on_rule["table_header"]
+        d["table_id"] = header.table_id
+        d["mip_era"] = header.mip_era
+        d["realm"] = " ".join(header.realm) if header.realm else header.realm
```

Comment:
this line makes no sense to me. Don't you want something like this:
```suggestion
        d["realm"] = " ".join(header.realm) if isinstance(header.realm, list) else header.realm
```